File: Algorithms/106. Construct Binary Tree from Inorder and Postorder Traversal/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        return self.f(inorder,postorder,0,len(inorder)-1,0,len(postorder)-1)
    def f(self, inorder, postorder, ib,ie,pb,pe):
        if ib>ie:
            return None
        val=postorder[pe]
        root=TreeNode(val)
        i=inorder.index(val)
        left=self.f(inorder,postorder,ib,i-1,pb,pb+(i-1-ib))
        right=self.f(inorder,postorder,i+1,ie,pb+(i-1-ib)+1,pe-1)
        root.left=left
        root.right=right
        return root

File: Algorithms/106. Construct Binary Tree from Inorder and Postorder Traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return build(inorder, postorder,0,inorder.length-1,0,postorder.length-1);
    }
    public TreeNode build(int[] inorder, int[] postorder,int is,int ie,int ps,int pe){
        if(ps>pe){
            return null;
        }
        int val=postorder[pe];
        TreeNode node=new TreeNode(val);
        int i = is;
        for (; i <=ie; i++) {
            if (inorder[i]==val) {
                break;
            }
        }
        int rang=i-is;
        node.left=build(inorder, postorder,is,i-1,ps,ps+rang-1);
        node.right=build(inorder,postorder,i+1,ie,ps+rang,pe-1);
        return node;
        
    }
}